pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -277,7 +277,19 @@ func (c *Reconciler) resolvePipelineState(
 	pst := resources.PipelineRunState{}
 	// Resolve each task individually because they each could have a different reference context (remote or local).
 	for _, task := range tasks {
-		fn, err := tresources.GetTaskFunc(ctx, c.KubeClientSet, c.PipelineClientSet, task.TaskRef, pr.Namespace, pr.Spec.ServiceAccountName)
+		// Create a dummy TaskRun for remote resolution.
```

Comment:
So this is what I'm most unsure about - we resolve the `Task` whether or not there's a `TaskRun` or even if we're creating a `TaskRun`, I believe for situations like when expressions and conditions. But the code in `pkg/remote` requires a `kmeta.OwnerRefable`. It only ever uses that `kmeta.OwnerRefable` to get its `Name` and `Namespace`, but nonetheless. Since we need to pass that `kmeta.OwnerRefable` in to `GetTaskFunc`, we don't have the `TaskRun`, if it exists, at the time we call `GetTaskFunc`.

So there's two aspects here - first, there's the `kmeta.OwnerRefable` stuff - that could be refactored to have the resolver code take name and namespace instead, and second, even if we did do that refactor, we'd still need to have the `TaskRun` name when calling `GetTaskFunc`, because https://github.com/tektoncd/pipeline/blob/adc127a5f1215019863768d58ad88bdf1a44fb5f/pkg/remote/resolution/resolver.go#L86 definitely wants the `TaskRun` name before resolving.

The options I see are the dummy `TaskRun` approach here, refactoring to take name and namespace instead, which still requires generating the `TaskRun` name at this location, or reworking the `PipelineRun` reconciler to _not_ resolve the `Task` until there's already a created `TaskRun`, deferring all the remote resolution logic to the `TaskRun` reconciler. The last option definitely feels like the ideal one for this particular change, but I'm not sure what the implications would be if we don't resolve the `Task` until we have a `TaskRun`.

Reply:
Ok, I'll poke around briefly to try to better understand (for my own edification, if nothing else!) when exactly we end up with a resolved `Task` but no `TaskRun` and what would be impacted by not resolving the `Task` unless/until we resolve the `TaskRun`, but so far, it does look like a lot of behavior, particularly around resolving task resources, would barf in that case. Still good to know better. =)

Aaaaah, yeah, you're right re the name. That said, `GetTaskRunName` is just checking to see if there already is a name defined in `pr.Status.ChildReferences` or `pr.Status.TaskRuns` for the pipeline task name, and if not, it returns `kmeta.ChildName(prName, fmt.Sprintf("-%s", ptName))`...which is just `prName-ptName` unless that would be longer than 63 characters (in which case it trims down `prName` and/or `ptName` and adds an md5 hash of the "full" string to guarantee uniqueness) anyway. So might as well just use `GetTaskRunName`.

Reply:
Aaah - I just checked deeper, and it turns out the `kmeta.OwnerRefable` ends up being used to add an owner reference to the generated `ResolutionRequest`:
https://github.com/tektoncd/resolution/blob/52bc1f17d4b6d8477e73a059de4b5808012a0386/pkg/resource/crd_resource.go#L104

So I think things need to be tweaked a bit more so that the `ResolutionRequest` has the `PipelineRun` as its owner reference.

Reply:
And we should definitely use `GetTaskRunName` for the remote resource name in `buildRequest` so that we ensure the resolution request made by the `PipelineRun` reconciler has the same unique identifier as the one made by the `TaskRun` reconciler to avoid an unnecessary repeat resolution request being created...yeah, I'd really prefer if we didn't try to resolve the `Task` from the `PipelineRun` reconciler at all. Grr.